### PR TITLE
Docker - support fat manifest

### PR DESCRIPTION
**Pull request by Or-Gabay**

### CODE REVIEW

File: artifactory/utils/docker/buildinfo.go [hunk 1]
```diff
@@ -316,7 +317,44 @@ func getConfigLayer(imageId string, searchResults map[string]utils.ResultItem, s
 }
 
 // Search for image layers in Artifactory.
-func searchImageLayers(imageId, imagePathPattern string, serviceManager *artifactory.ArtifactoryServicesManager) (map[string]utils.ResultItem, error) {
+func searchImageLayers(builder *buildInfoBuilder, imagePathPattern string, serviceManager *artifactory.ArtifactoryServicesManager) (map[string]utils.ResultItem, error) {
```
Comment (eyalbe4) [inline artifactory/utils/docker/buildinfo.go]: I suggest adding a unit test, which will checks the parsing of the fat manifest.

File: docker_test.go [hunk 2]
```diff
@@ -181,3 +211,25 @@ func TestDockerVersionScript(t *testing.T) {
 		t.Errorf("Could not get docker version, error: %s", err.Error())
 	}
 }
+
+func TestDockerFatManifestPull(t *testing.T) {
```
Comment (eyalbe4) [inline docker_test.go]: How is this test different from the existing TestDockerPull? Why is it needed?

File: artifactory/utils/docker/buildinfo.go [hunk 3]
```diff
@@ -316,7 +317,44 @@ func getConfigLayer(imageId string, searchResults map[string]utils.ResultItem, s
...
+
+	// Validate image ID layer exists.
+	if _, ok := resultMap[digestToLayer(builder.imageId)]; !ok {
+		// In case of a fat-manifest, Artifactory will create two folders.
+		// One folder named as the image tag, contain inside the fat manifest,
+		// The second folder, named as image's manifest digest, will contain the image layers and the image's manifest.
+		if _, ok := resultMap["list.manifest.json"]; ok {
+			v, _ := builder.image.Manifest()
+			var listManifest []Manifest
+			err := json.Unmarshal([]byte(v), &listManifest)
```
[the remark this answers is not in the document]
Comment (Or-Gabay) [reply artifactory/utils/docker/buildinfo.go]: Add log instead of new error

File: docker_test.go (hunk 2 above)
Comment (Or-Gabay) [reply docker_test.go]: Docker pull works with a non-fat-manifest image, unlike TestDockerFatManifestPull.
The idea behind this test is to validate the build info while pulling a fat-manifest-image from the remote repo.
The problem with remote repo is that Artifactory makes a different folder structure for pulling fat-manifest-images  witch break the collect build info step

File: artifactory/utils/docker/buildinfo.go (hunk 1 above)
Comment (Or-Gabay) [reply artifactory/utils/docker/buildinfo.go]: We don't parse but marshal the expected fat-manifest format,
TestDockerFatManifestPull should be all we need to check this

**Pull request merged**